js/classic/backend.js

```diff
@@ -388,17 +392,181 @@
     }
   };
 
-  backend.BluetoothDeviceTab = function(tabsManager, device) {
+  backend.BluetoothDeviceTab = function(tabsManager, adapter, device) {
     this._tabsManager = tabsManager;
+    this._adapter = adapter;
     this._device = device;
     this._tab = new ui.Tab(
         {onClose: this._onCloseHandler.bind(this)}, device.address);
 
+    this._attributesBlock =
+        new backend.BluetoothDeviceTab.AttributesBlock(this._tab, this._device);
+
+    this._pairBlock =
+        new backend.BluetoothDeviceTab.PairBlock(this._tab,
+                                                 this._adapter, this._device);
+    this._gattBlock = null;
+
+    this._attributeChangedHandler = this._generateAttributeChangedHandler();
+
     this._tabsManager.addTab(this._tab);
+
+    this._attributeChangedHandler({attrs: ['gatt']});
   };
 
   backend.BluetoothDeviceTab.prototype = {
     _onCloseHandler: function() {
+      this._attributesBlock.destroy();
+      this._pairBlock.destroy();
+      if (this._gattBlock !== null) {
+        this._gattBlock.destroy();
+      }
+
+      this._tabsManager.removeTab(this._tab);
+      this._tab.destroy();
+    },
+
+    _generateAttributeChangedHandler: function() {
+      var that = this;
+      return function(evt) {
+        evt.attrs.forEach(function(attrName) {
+          if (attrName == 'gatt') {
+            if (that._device.gatt !== null && that._gattBlock == null) {
+              that._gattBlock = new backend.BluetoothDeviceTab.GattBlock(
+                  that._tabsManager, that._tab, that._device);
+            } else if (that._device.gatt == null && that._gattBlock !== null) {
```

Comment:
I am not sure if `that._device.gatt == null` works or not. Because it is provided by Gecko.
I think `!that._device.gatt` would be safer. What do you think?